myst_nb/cache.py

```diff
@@ -0,0 +1,192 @@
+        jupyter_cache = env.config["jupyter_cache"]
+
+    if jupyter_cache is True:
+        path_cache = path_cache or Path(env.outdir).parent.joinpath(".jupyter_cache")
+        app.env.path_cache = (
+            path_cache  # TODO: is there a better way to make it accessible?
+        )
+
+        stage_and_execute(env, nb_list, path_cache)
+
+    return nb_list  # TODO: can also compare timestamps for inputs outputs
+
+
+def stage_and_execute(env, nb_list, path_cache):
+    pk_list = None
+    do_run = False
+
+    if path_cache:
+        try:
+            from jupyter_cache.cache.main import JupyterCacheBase  # noqa: F401
+        except ImportError:
+            logger.error(
+                (
+                    "Using caching functionality requires that "
+                    "jupyter_cache is uninstalled. Install it first."
+                )
+            )
+
+        cache_base = get_cache(path_cache)
+
+        if "jupyter_notebook_force_run" in env.config:
+            do_run = env.config["jupyter_notebook_force_run"]
+
+        for nb in nb_list:
+            # excludes the file with patterns given in execution_excludepatterns
+            # conf variable from executing, like index.rst
+            exclude_file = [x in nb for x in env.config["execution_excludepatterns"]]
+            if True in exclude_file:
+                continue
+
+            has_outputs = False
+
+            if "/" in nb:  # nb includes the path to notebook
+                source_path = nb
+            else:
+                source_path = env.env.doc2path(nb)
+
+            has_outputs = read_nb_output_cells(source_path, has_outputs, do_run)
+
+            if not has_outputs:
+                if pk_list is None:
+                    pk_list = []
+                stage_record = cache_base.stage_notebook_file(source_path)
+                pk_list.append(stage_record.pk)
+            else:
+                if has_outputs:  # directly cache the already executed notebooks
+                    cache_base.cache_notebook_file(source_path, overwrite=True)
+                logger.error(
+                    f"Will not run notebook with pre-populated outputs/no output cells, will directly cache: {source_path}"  # noqa:E501
+                )
+
+        execute_staged_nb(
+            cache_base, pk_list
+        )  # can leverage parallel execution implemented in jupyter-cache here
+
+
+def add_notebook_outputs(env, ntbk, file_path=None):
+    """
+    Add outputs to a NotebookNode by pulling from cache.
+
+    Function to get the database instance. Get the cached output of the notebook
+    and merge it with the original notebook. If there is no cached output,
+    checks if there was error during execution, then saves the traceback to a log file.
+    """
+    # If we have a jupyter_cache, see if there's a cache for this notebook
+    file_path = file_path or env.doc2path(env.docname)
+    dest_path = env.app.outdir
+    reports_dir = dest_path + "/reports"
```

Comment:
use Pathlib in case somebody is on windows

Reply:
you mean something like this ? :- 
```
dest_path = Path(env.app.outdir)
reports_dir = str(dest_path) + "/reports"
```